content/courses/advanced-ada/chapters/access_types.rst

```diff
@@ -294,14 +294,289 @@ Unchecked Access
+cannot use the :ada:`Free` procedure for the :ada:`Integer_Access` type when
+deallocating objects associated with the :ada:`Another_Integer_Access` type,
+even though both types are declared as :ada:`access Integer`.
+
+Note that we can use any name when instantiating the
+:ada:`Unchecked_Deallocation` procedure. However, naming it :ada:`Free` is very
+common.
+
+Now, let's see a complete example that includes object allocation and
+deallocation:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Unchecked_Deallocation
+
+    with Ada.Unchecked_Deallocation;
+
+    package Integer_Types is
+
+       type Integer_Access is access Integer;
+
+       procedure Free is
+         new Ada.Unchecked_Deallocation (Object => Integer,
+                                         Name   => Integer_Access);
+
+       procedure Show_Is_Null (I : Integer_Access);
+
+    end Integer_Types;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body Integer_Types is
+
+       procedure Show_Is_Null (I : Integer_Access) is
+       begin
+          if I = null then
+             Put_Line ("access value is null.");
+          else
+             Put_Line ("access value is NOT null.");
+          end if;
+       end Show_Is_Null;
+
+    end Integer_Types;
+
+    with Ada.Text_IO;   use Ada.Text_IO;
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I : Integer_Access;
+
+    begin
+       Put ("We haven't called new yet... ");
+       Show_Is_Null (I);
+
+       Put ("Calling new... ");
+       I := new Integer;
+       Show_Is_Null (I);
+
+       Put ("Calling Free... ");
+       Free (I);
+       Show_Is_Null (I);
+    end Show_Unchecked_Deallocation;
+
+In the :ada:`Show_Unchecked_Deallocation` procedure, we first allocate an
+object for :ada:`I` and then call :ada:`Free (I)` to deallocate it. Also, we
+call the :ada:`Show_Is_Null` procedure at three different points: before any
+allocation takes place, after allocating an object for :ada:`I`, and after
+deallocating that object.
+
+When we deallocate an object via a call to :ada:`Free`, the corresponding
+access value |mdash| which was previously pointing to an existing object
+|mdash| is set to :ada:`null`. Therefore, :ada:`I = null` after the call to
+:ada:`Free`, which is exactly what we see when running this example code.
+
+Note that it is OK to call :ada:`Free` for an access value that has already
+been deallocated:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Unchecked_Deallocation
+
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I : Integer_Access;
+
+    begin
+       I := new Integer;
+
+       Free (I);
+       Free (I);
+       Free (I);
+    end Show_Unchecked_Deallocation;
+
+The multiple calls to :ada:`Free` in this example don't cause any issues. In
+fact, calling :ada:`Free` to an access value that is equal to :ada:`null` is
+simply ignored.
```

Comment:
Do you want to talk about the danger of calling Free on the same designated object using another access value also designating that same object?

Reply:
I added a new subsection on dangling references (after this one) that discuss calls to `Free` when two access values are used.

Reply:
Just let me know if I should add another example.